File: 2019/python/day20.py
import heapq

neighbours = [(1, 0), (-1, 0), (0, -1), (0, 1)]
OUTER, INNER = 0, 1

def build_graph(m, start):
    portals = dict()
    w, h = len(m[0]), len(m)
    for y in range(2, h - 2):
        for x in range(2, w - 2):
            if m[y][x] != '.':
                continue

            if m[y][x-2:x].isalpha():
                name = m[y][x-2:x]
            elif m[y][x+1:x+3].isalpha():
                name = m[y][x+1:x+3]
            elif (m[y-2][x]+m[y-1][x]).isalpha():
                name = m[y-2][x]+m[y-1][x]
            elif (m[y+1][x]+m[y+2][x]).isalpha():
                name = m[y+1][x]+m[y+2][x]
            else:
                continue

            if name == start:
                start_pos = (x, y)
            elif x == 2 or y == 2 or x == w - 3 or y == h - 3:
                portals[x,y] = (name, OUTER)
            else:
                portals[x,y] = (name, INNER)

    g = {portal : find_adjacent(maze, portals, pos) for pos, portal in portals.items()}
    g[start] = find_adjacent(maze, portals, start_pos)
    return g

def find_adjacent(maze, portals, pos):
    visited = {pos}
    adjacent = []
    x, y = pos
    cells = [(x+dx, y+dy, 1) for (dx, dy) in neighbours if maze[y+dy][x+dx] == '.']

    while cells:
        (x, y, steps) = cells.pop(0)
        if (x, y) in visited:
            continue
        visited.add((x, y))

        if (x, y) in portals:
            adjacent.append((portals[x, y], steps))
        else:
            cells.extend([(x+dx, y+dy, steps + 1) for (dx, dy) in neighbours if maze[y+dy][x+dx] == '.'])
    return adjacent

def walk_maze(g, start, goal):
    visited = {start}
    q = [(dist, portal) for portal, dist in g[start]]
    heapq.heapify(q)

    while q:
        (dist, portal) = heapq.heappop(q)
        if portal[0] == goal:
            return dist
        if portal in visited:
            continue

        name, type = portal
        for next, d in g[(name, 1-type)]:
            heapq.heappush(q, (dist + d + 1, next))
        visited.add(portal)

def walk_maze_levels(g, start, goal):
    visited = {(start, 0)}
    q = [(dist, 0, portal) for portal, dist in g[start]]
    heapq.heapify(q)

    while q:
        (dist, lvl, portal) = heapq.heappop(q)
        if portal[0] == goal:
            if lvl == 0:
                return dist
            else:
                continue
        if (portal, lvl) in visited:
            continue

        name, type = portal
        for next, d in g[(name, 1-type)]:
            if type == INNER:
                heapq.heappush(q, (dist + d + 1, lvl + 1, next))
            elif lvl > 0:
                heapq.heappush(q, (dist + d + 1, lvl - 1, next))
        visited.add((portal, lvl))


with open('../day_20.txt', 'r') as f:
    maze = f.read().splitlines()

g = build_graph(maze, 'AA')
print('part 1:', walk_maze(g, 'AA', 'ZZ'))
print('part 2:', walk_maze_levels(g, 'AA', 'ZZ'))

File: 2019/go/day20.go
package main

import (
	"AdventOfCode/util"
	"container/heap"
	"fmt"
)

const (
	inner = iota
	outer
)

var neighbours = [4]vec2{
	vec2{1, 0},
	vec2{-1, 0},
	vec2{0, -1},
	vec2{0, 1},
}

type vec2 struct {
	x int
	y int
}

type portal struct {
	name string
	side int
}

type node struct {
	portal
	dist int
}

type graph map[portal][]node

type item struct {
	portal portal
	dist   int
	level  int
}

type queue []item

func (q queue) Len() int { return len(q) }

func (q queue) Less(i, j int) bool {
	return q[i].dist < q[j].dist
}

func (q queue) Swap(i, j int) {
	q[i], q[j] = q[j], q[i]
}

func (q *queue) Push(x interface{}) {
	i := x.(item)
	*q = append(*q, i)
}

func (q *queue) Pop() interface{} {
	old := *q
	n := len(old)
	i := old[n-1]
	*q = old[0 : n-1]
	return i
}

func buildGraph(maze [][]byte) graph {
	portals := make(map[vec2]portal)
	height, width := len(maze), len(maze[0])
	for y, row := range maze {
		for x, c := range row {
			if c != '.' {
				continue
			}
			for _, n := range neighbours {
				c0 := maze[y+n.y][x+n.x]
				if c0 < 'A' || c0 > 'Z' {
					continue
				}

				c1 := maze[y+n.y+n.y][x+n.x+n.x]
				var name string
				if n.y > 0 || n.x > 0 {
					name = string(c0) + string(c1)
				} else {
					name = string(c1) + string(c0)
				}

				var side int
				if x == 2 || y == 2 || x == width-3 || y == height-3 {
					side = outer
				}

				portals[vec2{x, y}] = portal{name, side}
				break
			}
		}
	}

	type cell struct {
		vec   vec2
		steps int
	}
	g := make(graph)
	for vec, portal := range portals {
		var adjacent []node
		var queue []cell
		for _, n := range neighbours {
			nx, ny := vec.x+n.x, vec.y+n.y
			if maze[ny][nx] == '.' {
				queue = append(queue, cell{vec: vec2{nx, ny}, steps: 1})
			}
		}

		visited := make(map[vec2]bool)
		visited[vec] = true

		for len(queue) > 0 {
			c := queue[0]
			queue = queue[1:]

			if visited[c.vec] {
				continue
			}
			visited[c.vec] = true

			if p, ok := portals[c.vec]; ok {
				adjacent = append(adjacent, node{portal: p, dist: c.steps})
				continue
			}

			for _, n := range neighbours {
				nx, ny := c.vec.x+n.x, c.vec.y+n.y
				if maze[ny][nx] == '.' {
					queue = append(queue, cell{vec: vec2{nx, ny}, steps: c.steps + 1})
				}
			}
		}
		g[portal] = adjacent
	}
	return g
}

func walkMaze(g graph, start string, end string, rec bool) int {
	type visit struct {
		name  string
		side  int
		level int
	}

	visited := make(map[visit]bool)
	var q queue
	for _, node := range g[portal{start, outer}] {
		q = append(q, item{portal: node.portal, dist: node.dist, level: 0})
	}
	heap.Init(&q)
	for len(q) > 0 {
		it := heap.Pop(&q).(item)
		p := it.portal
		if p.name == end && it.level == 0 {
			return it.dist
		}

		v := visit{p.name, p.side, it.level}
		if visited[v] {
			continue
		}
		visited[v] = true

		out := portal{p.name, 1 - p.side}
		for _, node := range g[out] {
			l := it.level
			if rec {
				if p.side == outer {
					l = it.level - 1
				} else {
					l = it.level + 1
				}
			}
			if l >= 0 {
				heap.Push(&q, item{portal: node.portal, dist: 1 + it.dist + node.dist, level: l})
			}
		}
	}
	return -1
}

func main() {
	input, _ := util.ReadInput("../day_20.txt")
	maze := make([][]byte, len(input))
	for i, l := range input {
		maze[i] = []byte(l)
	}
	g := buildGraph(maze)
	fmt.Printf("part 1: %d\n", walkMaze(g, "AA", "ZZ", false))
	fmt.Printf("part 2: %d\n", walkMaze(g, "AA", "ZZ", true))
}
